File: client/src/main/scala/stasis/client/api/clients/ServerApiEndpointClient.scala
package stasis.client.api.clients

import java.time.Instant

import scala.concurrent.Future

import org.apache.pekko.Done
import org.apache.pekko.util.ByteString

import stasis.client.model.DatasetMetadata
import stasis.core.commands.proto.Command
import stasis.core.discovery.ServiceApiClient
import stasis.shared.api.requests.CreateDatasetDefinition
import stasis.shared.api.requests.CreateDatasetEntry
import stasis.shared.api.requests.ResetUserPassword
import stasis.shared.api.requests.UpdateDatasetDefinition
import stasis.shared.api.responses.CreatedDatasetDefinition
import stasis.shared.api.responses.CreatedDatasetEntry
import stasis.shared.api.responses.Ping
import stasis.shared.api.responses.UpdatedUserSalt
import stasis.shared.model.datasets.DatasetDefinition
import stasis.shared.model.datasets.DatasetEntry
import stasis.shared.model.devices.Device
import stasis.shared.model.schedules.Schedule
import stasis.shared.model.users.User

trait ServerApiEndpointClient extends ServiceApiClient {
  def self: Device.Id
  def server: String

  def datasetDefinitions(): Future[Seq[DatasetDefinition]]
  def datasetDefinition(definition: DatasetDefinition.Id): Future[DatasetDefinition]
  def createDatasetDefinition(request: CreateDatasetDefinition): Future[CreatedDatasetDefinition]
  def updateDatasetDefinition(definition: DatasetDefinition.Id, request: UpdateDatasetDefinition): Future[Done]
  def deleteDatasetDefinition(definition: DatasetDefinition.Id): Future[Done]

  def datasetEntries(definition: DatasetDefinition.Id): Future[Seq[DatasetEntry]]
  def datasetEntry(entry: DatasetEntry.Id): Future[DatasetEntry]
  def latestEntry(definition: DatasetDefinition.Id, until: Option[Instant]): Future[Option[DatasetEntry]]
  def createDatasetEntry(request: CreateDatasetEntry): Future[CreatedDatasetEntry]
  def deleteDatasetEntry(entry: DatasetEntry.Id): Future[Done]

  def publicSchedules(): Future[Seq[Schedule]]
  def publicSchedule(schedule: Schedule.Id): Future[Schedule]

  def datasetMetadata(entry: DatasetEntry.Id): Future[DatasetMetadata]
  def datasetMetadata(entry: DatasetEntry): Future[DatasetMetadata]

  def user(): Future[User]
  def resetUserSalt(): Future[UpdatedUserSalt]
  def resetUserPassword(request: ResetUserPassword): Future[Done]

  def device(): Future[Device]
  def pushDeviceKey(key: ByteString): Future[Done]
  def pullDeviceKey(): Future[Option[ByteString]]
  def deviceKeyExists(): Future[Boolean]

  def ping(): Future[Ping]
  def commands(lastSequenceId: Option[Long]): Future[Seq[Command]]
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/api/clients/ServerApiEndpointClient.kt
package stasis.client_android.lib.api.clients

import okio.ByteString
import stasis.client_android.lib.discovery.ServiceApiClient
import stasis.client_android.lib.model.DatasetMetadata
import stasis.client_android.lib.model.server.api.requests.CreateDatasetDefinition
import stasis.client_android.lib.model.server.api.requests.CreateDatasetEntry
import stasis.client_android.lib.model.server.api.requests.ResetUserPassword
import stasis.client_android.lib.model.server.api.requests.UpdateDatasetDefinition
import stasis.client_android.lib.model.server.api.responses.CreatedDatasetDefinition
import stasis.client_android.lib.model.server.api.responses.CreatedDatasetEntry
import stasis.client_android.lib.model.server.api.responses.Ping
import stasis.client_android.lib.model.server.api.responses.UpdatedUserSalt
import stasis.client_android.lib.model.server.datasets.DatasetDefinition
import stasis.client_android.lib.model.server.datasets.DatasetDefinitionId
import stasis.client_android.lib.model.server.datasets.DatasetEntry
import stasis.client_android.lib.model.server.datasets.DatasetEntryId
import stasis.client_android.lib.model.server.devices.Device
import stasis.client_android.lib.model.server.devices.DeviceId
import stasis.client_android.lib.model.server.schedules.Schedule
import stasis.client_android.lib.model.server.schedules.ScheduleId
import stasis.client_android.lib.model.server.users.User
import stasis.client_android.lib.utils.Try
import stasis.core.commands.proto.Command
import java.time.Instant

@Suppress("TooManyFunctions")
interface ServerApiEndpointClient : ServiceApiClient {
    val self: DeviceId
    val server: String

    suspend fun datasetDefinitions(): Try<List<DatasetDefinition>>
    suspend fun datasetDefinition(definition: DatasetDefinitionId): Try<DatasetDefinition>
    suspend fun createDatasetDefinition(request: CreateDatasetDefinition): Try<CreatedDatasetDefinition>
    suspend fun updateDatasetDefinition(definition: DatasetDefinitionId, request: UpdateDatasetDefinition): Try<Unit>
    suspend fun deleteDatasetDefinition(definition: DatasetDefinitionId): Try<Unit>

    suspend fun datasetEntries(definition: DatasetDefinitionId): Try<List<DatasetEntry>>
    suspend fun datasetEntry(entry: DatasetEntryId): Try<DatasetEntry>
    suspend fun latestEntry(definition: DatasetDefinitionId, until: Instant?): Try<DatasetEntry?>
    suspend fun createDatasetEntry(request: CreateDatasetEntry): Try<CreatedDatasetEntry>
    suspend fun deleteDatasetEntry(entry: DatasetEntryId): Try<Unit>

    suspend fun publicSchedules(): Try<List<Schedule>>
    suspend fun publicSchedule(schedule: ScheduleId): Try<Schedule>

    suspend fun datasetMetadata(entry: DatasetEntryId): Try<DatasetMetadata>
    suspend fun datasetMetadata(entry: DatasetEntry): Try<DatasetMetadata>

    suspend fun user(): Try<User>
    suspend fun resetUserSalt(): Try<UpdatedUserSalt>
    suspend fun resetUserPassword(request: ResetUserPassword): Try<Unit>

    suspend fun device(): Try<Device>
    suspend fun pushDeviceKey(key: ByteString): Try<Unit>
    suspend fun pullDeviceKey(): Try<ByteString>
    suspend fun deviceKeyExists(): Try<Boolean>

    suspend fun ping(): Try<Ping>
    suspend fun commands(lastSequenceId: Long?): Try<List<Command>>
}
